services/spar/src/Spar/Scim.hs

```diff
@@ -108,8 +107,45 @@ apiScim = hoistScim (toServant (Scim.siteServer configuration))
      :<|> apiScimToken
   where
     hoistScim = hoistServer (Proxy @(Scim.SiteAPI ScimToken))
-                            (Scim.fromScimHandler fromError)
-    fromError = throwError . SAML.CustomServant . Scim.scimToServantErr
+                            (wrapScimErrors . toSpar)
+
+    -- Unwrap the 'ScimHandler'
+    toSpar :: Scim.ScimHandler Spar a -> Spar a
+    toSpar = Scim.fromScimHandler
+             (throwError . SAML.CustomServant . Scim.scimToServantErr)
+
+    -- Wrap all errors into the format required by SCIM.
+    --
+    -- FIXME: this doesn't catch impure exceptions (e.g. thrown with 'error').
+    -- Let's hope that SCIM clients can handle non-SCIM-formatted errors
+    -- properly. See <https://github.com/haskell-servant/servant/issues/1022>
+    -- for why it's hard to catch impure ex
+    wrapScimErrors :: Spar a -> Spar a
+    wrapScimErrors = over (_Spar . mapped) $ \io ->
+        try @_ @SomeException io <&> \case
+            -- We caught an exception that's not a Spar exception. It should be
+            -- wrapped into SCIM.serverError and rethrown.
+            Left someException ->
+                Left . SAML.CustomServant . Scim.scimToServantErr $
+                Scim.serverError (cs (displayException someException))
+            -- We caught a 'CustomServant' exception. It should be left as-is
+            -- because by now all SCIM errors have become 'CustomServant' errors
+            -- (thanks to 'toSpar') and we don't want to wrap them *again*.
+            Right (Left (SAML.CustomServant x)) ->
+                Left (SAML.CustomServant x)
+            -- We caught some other Spar exception. It should be wrapped into
+            -- SCIM.serverError and rethrown.
+            Right (Left sparError) ->
+                Left . SAML.CustomServant . Scim.scimToServantErr $
+                Scim.serverError (cs (errBody (sparToServantErr sparError)))
```

Comment:
Why not just `Left sparError`?  Then it will be converted later on in `sparToServantErr` like before.

Come to think of it, I would like to add a constructor `... | SparScimError ScimError` to `SparCustomError`, change `toSpar` as follows, and leave the conversion to `ServantErr` to the `Error` module:

```
toSpar = Scim.fromScimHandler (throwError . SparScimError)
```

We will still have this weird construct where the saml library wraps all errors, even those that have nothing to do with saml, but it keeps the errors in the app logic more strongly typed, and moves the burdon of rendering them more cleanly to the `Error` module.

Reply:
(do you want me to do that?)

Reply:
I'm okay with the constructor, as long as the /scim/v2 branch uses the SCIM format for all errors (as the spec requires). We also should take care to not wrap SCIM errors twice.

Reply:
agreed.  let me know if you want me to do any of this.

Reply:
I will get around to it at some point but I'm not sure when.